cmd/log/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Validate};
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list all stringbufs
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short, conflicts_with = "list")]
+    monitor: bool,
+    /// how long to delay in ms between polling for new log entries
+    #[clap(
+        long,
+        short,
+        default_value = "90",
+        requires = "monitor",
+        conflicts_with = "list"
+    )]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+fn stringbuf_monitor(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    delay: Option<u64>,
+    once: bool,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(ringbuf_var.size, Default::default);
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // Load the stringbuf struct with data from buf
+        let ringbuf_val: Value = Value::Struct(
+            reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+        );
+
+        let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+        Stringbuf::from_value(&cell.cell.value).unwrap()
+    };
+
+    //
+    // TODO
+    // set ctrl c handler before looping
+    // this will ensure that op_done is always called even if interrupted
+    // see jira https://rivosinc.atlassian.net/browse/SW-440
+    //
+    // represents the last seen values from the string buffer
+    let mut last_last: u32 = 0;
+    let mut last_buffer = vec![];
+    loop {
+        //
+        // Read the log from the core and convert to a u8 slice
+        //
+        let stringbuf = update_stringbuf();
+        let buffer = stringbuf.buffer.as_slice();
+        let last = stringbuf.last.unwrap();
+
+        // don't print duplicates
+        if last_last == last && last_buffer == stringbuf.buffer {
+            println!("skipped");
+            continue;
+        }
+
+        // read everything between current last, and last last by
+        // generating index range which connects these two in a circular manner, only including the
+        // new characters, and in the order they should be printed
+        // last_last has been printed already, last has not
+        let new_idxs: Vec<u32> = if last > last_last {
+            (last_last + 1..last).collect()
+
+        // This logic also works for last == last_last since we just need to print the entire range
```

Comment:
Does this comment apply to the if block or else block?

Reply:
The else.  I moved it inside the block to make it more clear.